src/com/xilinx/rapidwright/design/DesignTools.java

```diff
@@ -2345,7 +2395,128 @@ public static void copyImplementation(Design src, Design dest, boolean lockPlace
+			put("HMUX", Arrays.asList("H5LUT", "H6LUT"));
+		}};
+
+		// Map from a node to its driver PIP
+		// Note: Some PIPs are bidirectional. But, every PIP performs only one direction.
+		// The direction of a bidirectional PIP is determined from the context, ie., its connecting directional PIPs.
+		// To determine that context, go through directional PIPs first. This process does not support consecutive bidirectional PIPs.
+		Map<Net,Map<Node,PIP>> netToUphillPIPMap = new HashMap<>();
+		// netToPIPs to store PIPs extracted for static nets.
+		Map<Net,Set<PIP>> netToPIPs = new HashMap<>();
+		for (Net net : staticNets) {
+			netToPIPs.put(net, new HashSet<>());
+			Map<Node,PIP> nodeToDriverPIP = new HashMap<>();
+			List<PIP> biPIPs = new ArrayList<>();
+			// TODO: This goes through the whole design. Try to limit it to the target cell, if possible.
+			for (PIP pip : net.getPIPs()) {
+				if (pip.isBidirectional()) {
+					biPIPs.add(pip);
+				} else {
+					nodeToDriverPIP.put(pip.getEndNode(), pip);
+				}
+			}
+			for (PIP pip : biPIPs) {
+				Node stNode = pip.getStartNode();
+				if (nodeToDriverPIP.containsKey(stNode)) {
+					nodeToDriverPIP.put(pip.getEndNode(), pip);
+				} else {
+					nodeToDriverPIP.put(stNode, pip);
+				}
+			}
+			netToUphillPIPMap.put(net, nodeToDriverPIP);
+		}
+
+		Set<String> prohibitBels = new HashSet<>();
+
+		for (SiteInst siteInst : siteInstsOfCells) {
+			// Go through the set of pins being used on this site instance.
+			for (SitePinInst sitePinInst : siteInst.getSitePinInsts()) {
+				if (sitePinInst.isOutPin())
+					continue;
+
+				Net net = sitePinInst.getNet();
+				Map<Node,PIP> nodeToDriverPIP = netToUphillPIPMap.get(net);
+				if (nodeToDriverPIP != null) {
+					Set<PIP> allPIPs = netToPIPs.get(net);
+					// This SitePinInst connects to a static net. Trace and collect all the PIPs to a source.
+					Node node = sitePinInst.getConnectedNode();
+					SitePin sitePin = node.getSitePin();
+					// Backtrack through routing nodes (no SitePin)
+					while ((sitePin == null) || sitePin.isInput()) {
+						PIP pip = nodeToDriverPIP.get(node);
+						allPIPs.add(pip);
+						if (pip.isBidirectional()) {
+							node = pip.getStartNode().equals(node) ? pip.getEndNode() : pip.getStartNode();
+						} else {
+							node = pip.getStartNode();
+						}
+						if (node.getWireName().contains("VCC_WIRE"))
+							break;
+
+						sitePin = node.getSitePin();
+					}
+					if ((sitePin != null) && !node.getWireName().contains("VCC_WIRE"))  { // GND source
+						String  pinName = sitePin.getPinName();
+						String  siteName = sitePin.getSite().getName();
+						List<String> bels = sitePin2Bels.get(pinName);
+						for (String bel : bels) {
+							prohibitBels.add(siteName + "/" + bel);
+						}
+					}
+					netToPIPs.put(net,allPIPs);
+				}
+			}
+		}
+
+		for (String bel : prohibitBels) {
+			dest.addXDCConstraint(ConstraintGroup.LATE, "set_property PROHIBIT true [get_bels " + bel + "]");
+		}
+
```

Comment:
I also recall that we did or wanted to look at PROHIBIT-ing these BELs in a different way than using a XDC constraint. Do you remember what the outcome was?

Using XDC doesn't seem to be the best way if the consumer was RapidWright?

Reply:
Eddie, I remember that you want to use tag. IIRC, after your investigation, you conclude that it is not possible.

You have a point that this won't work if RapidWright placer is invoked after. But, I don't think we can have a good solution for that in a reasonable timeframe. I will add TODO for that.

Reply:
I have a faint recollection I/we looked at this but it's not clear why we wouldn't be able to block off those BELs.

Perhaps @clavin-xlnx can shed some light here?

Reply:
From what we looked at, Vivado's only mode to exclude placement on those locations is with the `PROHIBIT` constraint.  There is obviously the `EXCLUDE_PLACEMENT` on a pblock, but that is less precise.  

I don't exactly like the idea of adding the `PROHIBIT` constraints, but I don't have any better suggestions.

Reply:
We could use two modes. Prohibit a bel is to communicate to Vivado. Another is to communicate to downsteam RW (which we don't have and can be added when the need arise.)

Reply:
I was thinking of locked cells, set using `Cell.setLocked()` or `Cell.setEmptyLocked()`? What are the purposes of those cells if not to act as a prohibit?

Reply:
@clavin-xlnx What is your thought on this?     Are they transfered to dcp?  If not, they are not our options.

Reply:
Conclusion: `Cell.set*` methods don't achieve the same as `PROHIBIT` constraint.